File: mac/Carry/Engine/Container.swift
import Foundation

/// The iCloud container written by the Carry iOS app: health, location, inbox, manifest, license,
/// plus the heartbeat the Mac writes back so the phone can say "Mac read this 12 minutes ago" (container.py).
enum Container {
    static var dir: URL { CarryPaths.container }

    static func present() -> Bool { FileManager.default.fileExists(atPath: dir.path) }

    static func manifest() -> JSON? {
        let p = dir.appendingPathComponent("manifest.json")
        guard let data = try? Data(contentsOf: p) else { return nil }
        return JSON.parse(data)
    }

    /// `icloud_download(path)`: ask iCloud to materialise placeholders below `path`. Cheap, safe every sync.
    static func download(_ path: URL) {
        let fm = FileManager.default
        guard fm.fileExists(atPath: path.path) else { return }
        try? fm.startDownloadingUbiquitousItem(at: path)
        if let e = fm.enumerator(at: path, includingPropertiesForKeys: [.isUbiquitousItemKey], options: [.skipsHiddenFiles]) {
            for case let url as URL in e {
                if url.lastPathComponent.hasSuffix(".icloud") || url.lastPathComponent.hasPrefix(".") {
                    try? fm.startDownloadingUbiquitousItem(at: url)
                }
            }
        }
        // Placeholders are hidden files (".name.icloud"); a second pass without skipping catches them.
        if let e = fm.enumerator(at: path, includingPropertiesForKeys: nil) {
            for case let url as URL in e where url.pathExtension == "icloud" {
                try? fm.startDownloadingUbiquitousItem(at: url)
            }
        }
    }

    private static func jsonlFiles(_ sub: String) -> [URL] {
        Readers.glob(dir.appendingPathComponent(sub, isDirectory: true), suffix: ".jsonl")
    }

    /// Per-file byte offsets, stored as JSON in the cursor.
    private static func lineCursor(_ store: Store, _ key: String) -> [(String, JSON)] {
        guard let raw = try? store.getCursor(key), let j = JSON.parse(raw), case .object(let pairs) = j else { return [] }
        return pairs
    }

    private static func cursorValue(_ pairs: [(String, JSON)], _ key: String) -> Int {
        Int(pairs.first { $0.0 == key }?.1.int ?? 0)
    }

    private static func setCursorValue(_ pairs: inout [(String, JSON)], _ key: String, _ value: JSON) {
        if let i = pairs.firstIndex(where: { $0.0 == key }) { pairs[i].1 = value } else { pairs.append((key, value)) }
    }

    /// `_read_new_lines`: JSON objects after `offset`; if the file shrank it was replaced, start over.
    private static func readNewLines(_ path: URL, offset: Int) -> ([JSON], Int) {
        guard let data = try? Data(contentsOf: path) else { return ([], offset) }
        let from = offset > data.count ? 0 : offset
        var out: [JSON] = []
        let text = String(decoding: data[from...], as: UTF8.self)
        for line in Py.splitlines(text) {
            let trimmed = Py.strip(line)
            if trimmed.isEmpty { continue }
            if let j = JSON.parse(trimmed) { out.append(j) }
        }
        return (out, data.count)
    }

    /// `f"{v} {u}".strip()` for numbers, `f"{v}"` for strings.
    static func valueText(_ v: JSON?, _ u: String) -> String {
        let v = v ?? .null
        if case .string(let s) = v { return s }
        return Py.strip("\(v.pyStr) \(u)")
    }

    static func collectHealth(store: Store, cfg: JSON, backfillStart: Date) throws -> Int {
        download(dir.appendingPathComponent("health"))
        var cursors = lineCursor(store, "health")
        var new = 0
        for f in jsonlFiles("health") {
            let (recs, size) = readNewLines(f, offset: cursorValue(cursors, f.lastPathComponent))
            for r in recs {
                guard let start = PyTime.parseISO(r["start"]?.string) else { continue }
                let end = PyTime.parseISO(r["end"]?.string)
                let t = r["t"]?.string ?? "unknown"
                let u = r["u"]?.string ?? ""
                let src = r["src"]?.string ?? ""
                let text = valueText(r["v"], u)
                var meta: JSON = .object([("v", r["v"] ?? .null), ("u", .string(u)), ("src", .string(src))])
                if let extra = r["meta"], case .object = extra { meta.update(extra) }
                if try store.upsert(id: Util.stableID("health", t, r["start"], r["end"], src), source: "health", kind: t, ts: start,
                                    tsEnd: end, title: t, text: text, meta: meta, device: "iphone") {
                    new += 1
                }
            }
            setCursorValue(&cursors, f.lastPathComponent, .num(size))
        }
        try store.setCursor("health", JSON.object(cursors).dumps(), count: new)
        return new
    }

    static func collectLocation(store: Store, cfg: JSON, backfillStart: Date) throws -> Int {
        download(dir.appendingPathComponent("location"))
        var cursors = lineCursor(store, "location")
        var new = 0
        for f in jsonlFiles("location") {
            let (recs, size) = readNewLines(f, offset: cursorValue(cursors, f.lastPathComponent))
            for r in recs {
                let kind = r["kind"]?.string ?? "point"
                let ts: Date?
                let end: Date?
                if kind == "visit" {
                    ts = PyTime.parseISO(r["arrive"]?.string)
                    end = PyTime.parseISO(r["depart"]?.string)
                } else {
                    ts = PyTime.parseISO(r["ts"]?.string)
                    end = nil
                }
                guard let ts else { continue }
                let place = r["place"]
                let lat = r["lat"] ?? .null, lon = r["lon"] ?? .null
                let meta: JSON = .object([("lat", lat), ("lon", lon), ("acc_m", r["acc_m"] ?? .null), ("place", place ?? .null)])
                let when: JSON? = (r["arrive"]?.truthy == true ? r["arrive"] : nil) ?? r["ts"]
                let title = (place?.truthy == true ? place?.string : nil) ?? "\(lat.pyStr), \(lon.pyStr)"
                if try store.upsert(id: Util.stableID("loc", kind, when, lat, lon), source: "location", kind: kind, ts: ts, tsEnd: end,
                                    title: title, meta: meta, device: "iphone") {
                    new += 1
                }
            }
            setCursorValue(&cursors, f.lastPathComponent, .num(size))
        }
        try store.setCursor("location", JSON.object(cursors).dumps(), count: new)
        return new
    }

    static func collectInbox(store: Store, cfg: JSON, backfillStart: Date) throws -> Int {
        let d = dir.appendingPathComponent("inbox", isDirectory: true)
        download(d)
        guard FileManager.default.fileExists(atPath: d.path) else {
            try store.setCursor("inbox", nil, count: 0)
            return 0
        }
        var new = 0
        for f in Readers.glob(d, suffix: ".json") {
            guard let data = try? Data(contentsOf: f), let r = JSON.parse(data) else { continue }
            let ts = PyTime.parseISO(r["ts"]?.string) ?? PyTime.now()
            let kind = r["kind"]?.string ?? "text"
            let rawText = r["text"]?.string ?? ""
            let stripped = Py.strip(rawText)
            let text: String? = stripped.isEmpty ? nil : stripped // the note stays in meta; the digest prints it on its own line
            let fileName = r["file"]?.string
            let attachment = fileName.flatMap { $0.isEmpty ? nil : d.appendingPathComponent($0) }
            let meta: JSON = .object([("url", r["url"] ?? .null), ("from_app", r["from_app"] ?? .null), ("note", r["note"] ?? .null), ("file", r["file"] ?? .null)])
            let titleRaw = Py.strip(r["title"]?.string ?? "")
            let title = !titleRaw.isEmpty ? titleRaw
                : (r["url"]?.truthy == true ? r["url"]!.pyStr
                    : (Py.strip(Py.prefix(rawText, 60)).isEmpty ? Py.capitalize(kind) : Py.strip(Py.prefix(rawText, 60))))
            let idPart: Any? = (r["id"]?.truthy == true ? r["id"]?.pyStr : nil) ?? f.deletingPathExtension().lastPathComponent
            let path = attachment.flatMap { FileManager.default.fileExists(atPath: $0.path) ? $0.path : nil }
            if try store.upsert(id: Util.stableID("inbox", idPart), source: "inbox", kind: kind, ts: ts, title: title, text: text,
                                meta: meta, path: path, device: "iphone") {
                new += 1
            }
        }
        try store.setCursor("inbox", nil, count: new)
        return new
    }

    /// `heartbeat/<host>.json`, the same host rule as the CLI (`socket.gethostname()` before the first dot).
    static var hostName: String {
        var buffer = [CChar](repeating: 0, count: 256)
        gethostname(&buffer, buffer.count - 1)
        let full = String(cString: buffer)
        return full.split(separator: ".").first.map(String.init) ?? full
    }

    @discardableResult
    static func writeHeartbeat(store: Store, version: String, pro: Bool, sourcesAppliedAt: String?, lastDigestDate: String?) throws -> URL? {
        guard present() else { return nil }
        let d = dir.appendingPathComponent("heartbeat", isDirectory: true)
        try FileManager.default.createDirectory(at: d, withIntermediateDirectories: true)
        let host = hostName
        let counts = try store.counts()
        let payload: JSON = .object([
            ("host", .string(host)),
            ("carry_version", .string(version)),
            ("last_sync_at", .string(PyTime.iso(PyTime.now()))),
            ("last_digest_date", .str(lastDigestDate)),
            ("counts", .object(["health", "location", "inbox", "photos", "screenshots", "notes", "voice_memos"].map { ($0, .num(counts[$0] ?? 0)) })),
            ("pro", .bool(pro)),
            ("sources_applied_at", .str(sourcesAppliedAt)),
        ])
        let p = d.appendingPathComponent("\(host).json")
        let tmp = d.appendingPathComponent("\(host).json.tmp")
        try payload.dumps(indent: 2).write(to: tmp, atomically: false, encoding: .utf8)
        _ = try FileManager.default.replaceItemAt(p, withItemAt: tmp)
        return p
    }

    // MARK: - Health Auto Export fallback
    // Some people already run Health Auto Export (HealthyApps) with an iCloud Drive automation. Its JSON lands in
    // iCloud Drive › AutoExport › <automation>/ . Reading it is a fallback for when the Carry app's own HealthKit
    // channel is not available (unsigned build, App Review outcome, or the user simply prefers HAE).

    static let haeDirs: [URL] = [
        Readers.home.appendingPathComponent("Library/Mobile Documents/com~apple~CloudDocs/AutoExport", isDirectory: true),
    ]

    static let haeNames: [String: (String, String?)] = [
        "step_count": ("steps", "count"), "heart_rate": ("heart_rate", "bpm"), "resting_heart_rate": ("resting_heart_rate", "bpm"),
        "heart_rate_variability": ("hrv", "ms"), "active_energy": ("active_energy", "kcal"), "weight_body_mass": ("body_mass", nil),
        "blood_oxygen_saturation": ("blood_oxygen", "ratio"),
    ]

    static func haeDirectories(_ cfg: JSON) -> [URL] {
        var dirs: [URL] = []
        if let extra = cfg["carry"]?["health_auto_export_dir"]?.string, !extra.isEmpty {
            dirs.append(URL(fileURLWithPath: (extra as NSString).expandingTildeInPath, isDirectory: true))
        }
        dirs += haeDirs
        let mobile = Readers.home.appendingPathComponent("Library/Mobile Documents", isDirectory: true)
        if let names = try? FileManager.default.contentsOfDirectory(atPath: mobile.path) {
            for n in names where n.contains("HealthAutoExport") {
                dirs.append(mobile.appendingPathComponent(n).appendingPathComponent("Documents", isDirectory: true))
            }
        }
        return dirs.filter { FileManager.default.fileExists(atPath: $0.path) }
    }

    private static func haeDate(_ v: JSON?) -> Date? {
        guard let v else { return nil }
        let s = v.pyStr
        return PyTime.parseHAE(s)
    }

    /// `hae_records(payload)`: one Health Auto Export JSON document → DATA-CONTRACT §4 records.
    static func haeRecords(_ payload: JSON) -> [JSON] {
        var out: [JSON] = []
        let data = payload["data"] ?? payload
        func rec(_ t: String, _ start: Date, _ end: Date, _ v: JSON, _ u: String, _ src: String, meta: JSON? = nil) -> JSON {
            var pairs: [(String, JSON)] = [("t", .string(t)), ("start", .string(PyTime.iso(start))), ("end", .string(PyTime.iso(end))),
                                           ("v", v), ("u", .string(u)), ("src", .string(src))]
            if let meta { pairs.append(("meta", meta)) }
            return .object(pairs)
        }
        for metric in data["metrics"]?.array ?? [] {
            let name = metric["name"]?.string
            let units = metric["units"]?.string ?? ""
            for pt in metric["data"]?.array ?? [] {
                guard let when = haeDate(pt["date"] ?? .string("")) else { continue }
                let src = (pt["source"]?.truthy == true ? pt["source"]?.string : nil) ?? "Health Auto Export"
                if name == "sleep_analysis" {
                    let start = haeDate((pt["sleepStart"]?.truthy == true ? pt["sleepStart"] : nil) ?? (pt["inBedStart"]?.truthy == true ? pt["inBedStart"] : nil) ?? .string("")) ?? when
                    var cursor = start
                    for (stage, key) in [("core", "core"), ("deep", "deep"), ("rem", "rem"), ("awake", "awake")] {
                        guard let hours = pt[key], hours.truthy, let h = hours.double else { continue }
                        let end = cursor.addingTimeInterval(h * 3600)
                        out.append(rec("sleep", cursor, end, .string(stage), "stage", src, meta: .object([("approx", .bool(true))])))
                        cursor = end
                    }
                    let anyStage = ["core", "deep", "rem"].contains { pt[$0]?.truthy == true }
                    if !anyStage, let asleep = pt["asleep"], asleep.truthy, let h = asleep.double {
                        out.append(rec("sleep", start, start.addingTimeInterval(h * 3600), .string("asleepUnspecified"), "stage", src,
                                       meta: .object([("approx", .bool(true))])))
                    }
                    continue
                }
                guard let name, let (t, unit) = haeNames[name] else { continue }
                var value: JSON? = name == "heart_rate" ? (pt["Avg"] ?? pt["qty"]) : pt["qty"]
                if value == nil || value?.isNull == true { continue }
                if name == "blood_oxygen_saturation", let d = value?.double, d > 1.5 { value = .double(d / 100.0) }
                out.append(rec(t, when, when, value ?? .null, unit ?? units, src))
            }
        }
        for w in data["workouts"]?.array ?? [] {
            guard let start = haeDate(w["start"] ?? .string("")) else { continue }
            let end = haeDate(w["end"] ?? .string(""))
            var meta: JSON = .object([])
            if let dist = w["distance"], case .object = dist, let qty = dist["qty"], !qty.isNull, let q = qty.double {
                let units = (dist["units"]?.pyStr ?? "").lowercased()
                let factor: Double = ["km", "kilometer", "kilometers"].contains(units) ? 1000 : 1
                meta.set("distance_m", .double(q * factor))
            }
            if let energy = w["activeEnergyBurned"] ?? w["activeEnergy"], case .object = energy, let qty = energy["qty"], !qty.isNull, let q = qty.double {
                meta.set("energy_kcal", .double(q))
            }
            let name = (w["name"]?.truthy == true ? w["name"]?.pyStr : nil) ?? "workout"
            let src = (w["source"]?.truthy == true ? w["source"]?.pyStr : nil) ?? "Health Auto Export"
            out.append(rec("workout", start, end ?? start, .string(name), "type", src, meta: meta))
        }
        return out
    }

    static func collectHealthAutoExport(store: Store, cfg: JSON, backfillStart: Date) throws -> Int {
        var new = 0
        var cursors = lineCursor(store, "health_hae")
        let fm = FileManager.default
        for d in haeDirectories(cfg) {
            download(d)
            // Keys are the paths as configured (Python's `str(f)`), symlinks left alone: /tmp stays /tmp.
            var files: [URL] = []
            if let e = fm.enumerator(atPath: d.path) {
                for case let rel as String in e where rel.hasSuffix(".json") {
                    files.append(URL(fileURLWithPath: d.path + "/" + rel))
                }
            }
            files.sort { $0.path.utf8.lexicographicallyPrecedes($1.path.utf8) }
            for f in files {
                let key = f.path
                guard let attrs = try? fm.attributesOfItem(atPath: f.path) else { continue }
                let mtime = Int((attrs[.modificationDate] as? Date)?.timeIntervalSince1970 ?? 0)
                let size = (attrs[.size] as? NSNumber)?.intValue ?? 0
                let stamp = "\(mtime):\(size)"
                if cursors.first(where: { $0.0 == key })?.1.string == stamp { continue }
                guard let data = try? Data(contentsOf: f), let payload = JSON.parse(data) else { continue }
                for r in haeRecords(payload) {
                    guard let start = PyTime.parseISO(r["start"]?.string), start >= backfillStart else { continue }
                    let end = PyTime.parseISO(r["end"]?.string)
                    let u = r["u"]?.string ?? ""
                    let src = r["src"]?.string ?? ""
                    let t = r["t"]?.string ?? ""
                    let text = valueText(r["v"], u)
                    var meta: JSON = .object([("v", r["v"] ?? .null), ("u", .string(u)), ("src", .string(src)), ("via", .string("health-auto-export"))])
                    if let extra = r["meta"] { meta.update(extra) }
                    if try store.upsert(id: Util.stableID("health", t, r["start"], r["end"], src), source: "health", kind: t, ts: start,
                                        tsEnd: end, title: t, text: text, meta: meta, device: "iphone") {
                        new += 1
                    }
                }
                setCursorValue(&cursors, key, .string(stamp))
            }
        }
        try store.setCursor("health_hae", JSON.object(cursors).dumps(), count: new)
        return new
    }
}

File: cli/src/carry/container.py
"""The iCloud container written by the Carry iOS app: health, location, inbox, manifest, license,
plus the heartbeat the Mac writes back so the phone can say "Mac read this 12 minutes ago"."""

from __future__ import annotations

import json
import socket
from datetime import datetime, timedelta
from pathlib import Path

from carry import __version__
from carry.config import CONTAINER_DIR
from carry.util import icloud_download, iso, now, parse_iso, stable_id


def present() -> bool:
    return CONTAINER_DIR.exists()


def manifest() -> dict | None:
    p = CONTAINER_DIR / "manifest.json"
    try:
        return json.loads(p.read_text()) if p.exists() else None
    except Exception:  # noqa: BLE001
        return None


def _jsonl_files(sub: str) -> list[Path]:
    d = CONTAINER_DIR / sub
    return sorted(d.glob("*.jsonl")) if d.exists() else []


def _line_cursor(store, key: str) -> dict[str, int]:
    raw = store.get_cursor(key)
    try:
        return json.loads(raw) if raw else {}
    except Exception:  # noqa: BLE001
        return {}


def _read_new_lines(path: Path, offset: int) -> tuple[list[dict], int]:
    data = path.read_bytes()
    if offset > len(data):  # file was replaced; start over
        offset = 0
    out = []
    for line in data[offset:].splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            out.append(json.loads(line))
        except Exception:  # noqa: BLE001
            continue
    return out, len(data)


def collect_health(store, cfg, backfill_start: datetime) -> int:
    icloud_download(CONTAINER_DIR / "health")
    cursors = _line_cursor(store, "health")
    new = 0
    for f in _jsonl_files("health"):
        recs, size = _read_new_lines(f, cursors.get(f.name, 0))
        for r in recs:
            start, end = parse_iso(r.get("start", "")), parse_iso(r.get("end", ""))
            if not start:
                continue
            t = r.get("t", "unknown")
            v, u, src = r.get("v"), r.get("u", ""), r.get("src", "")
            text = f"{v} {u}".strip() if not isinstance(v, str) else f"{v}"
            meta = {"v": v, "u": u, "src": src}
            if isinstance(r.get("meta"), dict):
                meta.update(r["meta"])
            if store.upsert(id=stable_id("health", t, r.get("start"), r.get("end"), src), source="health", kind=t,
                            ts=start, ts_end=end, title=t, text=text, meta=meta, device="iphone"):
                new += 1
        cursors[f.name] = size
    store.set_cursor("health", json.dumps(cursors), new)
    return new


def collect_location(store, cfg, backfill_start: datetime) -> int:
    icloud_download(CONTAINER_DIR / "location")
    cursors = _line_cursor(store, "location")
    new = 0
    for f in _jsonl_files("location"):
        recs, size = _read_new_lines(f, cursors.get(f.name, 0))
        for r in recs:
            kind = r.get("kind", "point")
            if kind == "visit":
                ts, end = parse_iso(r.get("arrive", "")), parse_iso(r.get("depart", ""))
            else:
                ts, end = parse_iso(r.get("ts", "")), None
            if not ts:
                continue
            place = r.get("place")
            meta = {"lat": r.get("lat"), "lon": r.get("lon"), "acc_m": r.get("acc_m"), "place": place}
            if store.upsert(id=stable_id("loc", kind, r.get("arrive") or r.get("ts"), r.get("lat"), r.get("lon")),
                            source="location", kind=kind, ts=ts, ts_end=end, title=place or f"{r.get('lat')}, {r.get('lon')}",
                            meta=meta, device="iphone"):
                new += 1
        cursors[f.name] = size
    store.set_cursor("location", json.dumps(cursors), new)
    return new


def collect_inbox(store, cfg, backfill_start: datetime) -> int:
    d = CONTAINER_DIR / "inbox"
    icloud_download(d)
    if not d.exists():
        store.set_cursor("inbox", None, 0)
        return 0
    new = 0
    for f in sorted(d.glob("*.json")):
        try:
            r = json.loads(f.read_text())
        except Exception:  # noqa: BLE001
            continue
        ts = parse_iso(r.get("ts", "")) or now()
        kind = r.get("kind", "text")
        text = (r.get("text") or "").strip() or None  # the note stays in meta; the digest prints it on its own line
        attachment = d / r["file"] if r.get("file") else None
        meta = {"url": r.get("url"), "from_app": r.get("from_app"), "note": r.get("note"), "file": r.get("file")}
        title = (r.get("title") or "").strip() or r.get("url") or ((r.get("text") or "")[:60].strip()) or kind.capitalize()
        if store.upsert(id=stable_id("inbox", r.get("id") or f.stem), source="inbox", kind=kind, ts=ts, title=title,
                        text=text, meta=meta, path=str(attachment) if attachment and attachment.exists() else None,
                        device="iphone"):
            new += 1
    store.set_cursor("inbox", None, new)
    return new


def write_heartbeat(store, pro: bool, sources_applied_at: str | None, last_digest_date: str | None) -> Path | None:
    if not present():
        return None
    d = CONTAINER_DIR / "heartbeat"
    d.mkdir(parents=True, exist_ok=True)
    host = socket.gethostname().split(".")[0]
    counts = store.counts()
    payload = {
        "host": host,
        "carry_version": __version__,
        "last_sync_at": iso(now()),
        "last_digest_date": last_digest_date,
        "counts": {k: counts.get(k, 0) for k in ("health", "location", "inbox", "photos", "screenshots", "notes", "voice_memos")},
        "pro": pro,
        "sources_applied_at": sources_applied_at,
    }
    p = d / f"{host}.json"
    tmp = p.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(payload, ensure_ascii=False, indent=2))
    tmp.replace(p)
    return p


# --- Health Auto Export fallback --------------------------------------------------------------------
# Some people already run Health Auto Export (HealthyApps) with an iCloud Drive automation. Its JSON lands in
# iCloud Drive › AutoExport › <automation>/ . Reading it is a fallback for when the Carry app's own HealthKit
# channel is not available (unsigned build, App Review outcome, or the user simply prefers HAE).

HAE_DIRS = [
    Path.home() / "Library/Mobile Documents/com~apple~CloudDocs/AutoExport",
]
HAE_NAMES = {
    "step_count": ("steps", "count"), "heart_rate": ("heart_rate", "bpm"), "resting_heart_rate": ("resting_heart_rate", "bpm"),
    "heart_rate_variability": ("hrv", "ms"), "active_energy": ("active_energy", "kcal"), "weight_body_mass": ("body_mass", None),
    "blood_oxygen_saturation": ("blood_oxygen", "ratio"),
}


def hae_dirs(cfg) -> list[Path]:
    extra = (cfg.get("carry") or {}).get("health_auto_export_dir")
    dirs = [Path(extra).expanduser()] if extra else []
    dirs += HAE_DIRS
    dirs += list((Path.home() / "Library/Mobile Documents").glob("*HealthAutoExport*/Documents"))
    return [d for d in dirs if d.exists()]


def _hae_date(s: str):
    """'2024-02-06 14:30:00 -0800' → aware datetime."""
    try:
        return datetime.strptime(s, "%Y-%m-%d %H:%M:%S %z")
    except Exception:  # noqa: BLE001
        return parse_iso(s)


def hae_records(payload: dict) -> list[dict]:
    """Translate one Health Auto Export JSON document into DATA-CONTRACT §4 records."""
    out = []
    data = payload.get("data", payload)
    for metric in data.get("metrics", []) or []:
        name, units = metric.get("name"), metric.get("units") or ""
        for pt in metric.get("data", []) or []:
            when = _hae_date(str(pt.get("date", "")))
            if not when:
                continue
            src = pt.get("source") or "Health Auto Export"
            if name == "sleep_analysis":
                start = _hae_date(str(pt.get("sleepStart") or pt.get("inBedStart") or "")) or when
                cursor = start
                for stage, hours in (("core", pt.get("core")), ("deep", pt.get("deep")), ("rem", pt.get("rem")), ("awake", pt.get("awake"))):
                    if not hours:
                        continue
                    end = cursor + timedelta(hours=float(hours))
                    out.append({"t": "sleep", "start": iso(cursor), "end": iso(end), "v": stage, "u": "stage", "src": src,
                                "meta": {"approx": True}})
                    cursor = end
                if not any(pt.get(k) for k in ("core", "deep", "rem")) and pt.get("asleep"):
                    out.append({"t": "sleep", "start": iso(start), "end": iso(start + timedelta(hours=float(pt["asleep"]))),
                                "v": "asleepUnspecified", "u": "stage", "src": src, "meta": {"approx": True}})
                continue
            if name not in HAE_NAMES:
                continue
            t, unit = HAE_NAMES[name]
            if name == "heart_rate":
                value = pt.get("Avg", pt.get("qty"))
            else:
                value = pt.get("qty")
            if value is None:
                continue
            if name == "blood_oxygen_saturation" and value > 1.5:
                value = value / 100.0
            out.append({"t": t, "start": iso(when), "end": iso(when), "v": value, "u": unit or units, "src": src})
    for w in data.get("workouts", []) or []:
        start, end = _hae_date(str(w.get("start", ""))), _hae_date(str(w.get("end", "")))
        if not start:
            continue
        meta = {}
        dist = w.get("distance") or {}
        if isinstance(dist, dict) and dist.get("qty") is not None:
            meta["distance_m"] = float(dist["qty"]) * (1000 if str(dist.get("units", "")).lower() in ("km", "kilometer", "kilometers") else 1)
        energy = w.get("activeEnergyBurned") or w.get("activeEnergy") or {}
        if isinstance(energy, dict) and energy.get("qty") is not None:
            meta["energy_kcal"] = float(energy["qty"])
        out.append({"t": "workout", "start": iso(start), "end": iso(end or start), "v": w.get("name") or "workout", "u": "type",
                    "src": w.get("source") or "Health Auto Export", "meta": meta})
    return out


def collect_health_auto_export(store, cfg, backfill_start: datetime) -> int:
    new = 0
    cursors = _line_cursor(store, "health_hae")
    for d in hae_dirs(cfg):
        icloud_download(d)
        for f in sorted(d.rglob("*.json")):
            key = str(f)
            stamp = f"{int(f.stat().st_mtime)}:{f.stat().st_size}"
            if cursors.get(key) == stamp:
                continue
            try:
                payload = json.loads(f.read_text())
            except Exception:  # noqa: BLE001
                continue
            for r in hae_records(payload):
                start, end = parse_iso(r["start"]), parse_iso(r["end"])
                if not start or start < backfill_start:
                    continue
                text = f"{r['v']} {r['u']}".strip() if not isinstance(r["v"], str) else f"{r['v']}"
                meta = {"v": r["v"], "u": r["u"], "src": r["src"], "via": "health-auto-export"}
                meta.update(r.get("meta") or {})
                if store.upsert(id=stable_id("health", r["t"], r["start"], r["end"], r["src"]), source="health", kind=r["t"],
                                ts=start, ts_end=end, title=r["t"], text=text, meta=meta, device="iphone"):
                    new += 1
            cursors[key] = stamp
    store.set_cursor("health_hae", json.dumps(cursors), new)
    return new
